### DISCUSSION TITLE
Consumption of Nuget package outside of Visual Studio

### DISCUSSION THREAD

**Discussion by ordinaryorange** [cefsharp/CefSharp#3813] [General]
Apologies if I'm not supposed to post this here, but I did notice the [Questions-and-Support](https://github.com/cefsharp/Questions-and-Support/issues) has been archived.

I'm attempting to use CefSharp from [Linqpad](https://www.linqpad.net/)
Linqpad6 (64bit .net core version) and CefSharp.Offscreen.NetCore specifically.

`var b = new ChromiumWebBrowser();`
is all that is in the script.

Which results in 
**_BadImageFormatException_**
_Could not load file or assembly 'CefSharp.Core.Runtime, Version=93.1.111.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138'. An attempt was made to load a program with an incorrect format._

I can run CefSharp in Visual Studio without error.

I've tried all manner of hacking to get it to work including

- x86 version of Linqpad
- Manually referencing the dll's, [according to the docs](https://github.com/cefsharp/CefSharp/wiki/Output-files-description-table-(Redistribution)), rather than using a nuget package
- Wrapping everything in a visual studio published lib and then consuming that from LP

I suspect it either has something to do with how LP loads the dll's vs how VS does it, or CefSharp is doing some runtime dynamic assembly loading under the hood incorrectly?
I notice that there isn't even a 64bit version of CefSharp.Offscreen.dll in the nuget package, and I suspect this is only supposed to be a design time reference? But I'm over my head here.

Does this look like a CefSharp or Linqpad problem ?

**Comment by amaitland** [comment]
> Apologies if I'm not supposed to post this here,

Here in the discussions section is perfectly fine 👍 

> Which results in
> **_BadImageFormatException_**
> _Could not load file or assembly 'CefSharp.Core.Runtime, Version=93.1.111.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138'. An attempt was made to load a program with an incorrect format._

For `.Net Core/.Net 5.0` this can be misleading as detailed in https://github.com/cefsharp/CefSharp/issues/3284#issuecomment-805342115

It probably means it was unable to load one of the dependencies, I'm guessing.

> I suspect it either has something to do with how LP loads the dll's vs how VS does it,

Without knowing how `LinqPad` works it's hard to say.

> CefSharp is doing some runtime dynamic assembly loading under the hood incorrectly?

The default handling is used first. Basically all the unmanaged resources should be loaded relatively to `CefSharp.Core.Runtime.dll`

> I notice that there isn't even a 64bit version of CefSharp.Offscreen.dll in the nuget package, and I suspect this is only supposed to be a design time reference?

There is a [ref](https://docs.microsoft.com/en-us/dotnet/standard/assembly/reference-assemblies) assembly and a `MISL` targeted lib. There is no need for a `x64` bit version of `CefSharp.OffScreen.dll` specifically. Only `CefSharp.Core.Runtime.dll` has a `x64/x86/arm64` version.

**Reply by amaitland** [reply]
> Does this look like a CefSharp or Linqpad problem ?

Unless you'd like to sponsor my time this isn't something I have time to look into to workout where the problem lies.

**Reply by amaitland** [reply]
Should you wish to debug further you might find [Process Monitor](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) useful, you can look at the calls that are made to load the dlls, you might get a better understanding of where it's trying to load the dlls from without actually having to build  `Linqpad` from source.

**Reply by ordinaryorange** [reply]
Thanks for your feedback. I've been down the Process Monitor route and did not discover any nuggets.
Running in Linqpad is a "nice to have" not a "must have" so it is not worth consuming effort over.
If I keep persisting and discover a solution I'll post back for for completeness.

**Comment by ordinaryorange** [comment]
For anyone landing here, we have a solution - https://stackoverflow.com/questions/69923244/cefsharp-offscreen-in-linqpad/69997723#69997723
